### PR TITLE
[DH-3] let's tune us some tcp kernel settings!

### PR DESCRIPTION
**Pull request by shaneknapp**
addresses:  https://jira-secure.berkeley.edu/browse/DH-3

from the yaml:
```
   # Figures below are in bytes, and gleaned from the following articles:
   # https://www.ibm.com/docs/en/linux-on-systems?topic=tuning-tcpip-ipv4-settings
   # https://www.ibm.com/docs/en/linux-on-systems?topic=tuning-network-stack-settings
```

i also created a `vendor/` subdir for configs like this and am open to brainstorming naming/path conventions.

### CODE REVIEW

**Comment by shaneknapp** [comment]
@ryanlovett : 
image

you say that 1638400 bytes is 6G, but it's actually 1.6M.  my setting is 16M, based on this:  https://www.ibm.com/docs/en/linux-on-systems?topic=tuning-tcpip-ipv4-settings

i also set `net.core.rmem_max` and `wmem_max` to be double (32M) the values of `tcp_rmem` and `tcp_wmem`.  waves hands etc

**Comment by ryanlovett** [comment]
> you say that 1638400 bytes is 6G, but it's actually 1.6M. my setting is 16M, based on this: https://www.ibm.com/docs/en/linux-on-systems?topic=tuning-tcpip-ipv4-settings

I read in one of those articles that the unit is in pages where 1 page is 4096 bytes, i.e.
> The minimum represents the smallest receive buffer size guaranteed, even under memory pressure. The minimum value defaults to 1 page or 4096 bytes.

So 1638400 * 4096 = 6710886400. But I haven't checked what are the official units of that parameter.

**Comment by shaneknapp** [comment]
> > you say that 1638400 bytes is 6G, but it's actually 1.6M. my setting is 16M, based on this: https://www.ibm.com/docs/en/linux-on-systems?topic=tuning-tcpip-ipv4-settings
> 
> I read in one of those articles that the unit is in pages where 1 page is 4096 bytes, i.e.
> 
> > The minimum represents the smallest receive buffer size guaranteed, even under memory pressure. The minimum value defaults to 1 page or 4096 bytes.
> 
> So 1638400 * 4096 = 6710886400. But I haven't checked what are the official units of that parameter.

i think you're right.  looks like some of the vars are pages vs bytes.  i'll go over this tomorrow and commit some updates.

**Comment by shaneknapp** [comment]
@ryanlovett here's the command i propose running to create a new node pool w/these settings:
```
gcloud container node-pools create "core-2023-04-18"  \
  --machine-type "n2-highmem-8"  \
  --enable-autoscaling --min-nodes "0" --max-nodes "3" \
  --project "ucb-datahub-2018" --cluster "fall-2019" --region "us-central1" --node-locations "us-central1-b" \
  --tags hub-cluster \
  --image-type "COS_CONTAINERD" --disk-type "pd-balanced" --disk-size "200"  \
  --metadata disable-legacy-endpoints=true \
  --scopes "https://www.googleapis.com/auth/devstorage.read_only","https://www.googleapis.com/auth/logging.write","https://www.googleapis.com/auth/monitoring","https://www.googleapis.com/auth/servicecontrol","https://www.googleapis.com/auth/service.management.readonly","https://www.googleapis.com/auth/trace.append" \
  --no-enable-autoupgrade --enable-autorepair --max-surge-upgrade 1 --max-unavailable-upgrade 0 --max-pods-per-node "110" \
  --system-config-from-file=vendor/google/gke/node-pool/config/core-pool-sysctl.yaml
```
i removed all of the labels, k8s labels, taints, etc just to ensure that we don't have any unintended consequences etc.

**Review by ryanlovett — COMMENTED** [review]
It's great this is being added!

Wouldn't the command you proposed [also need](https://cloud.google.com/kubernetes-engine/docs/how-to/node-system-config#create-node-pool) `--system-config-from-file=vendor/google/gke/node-pool/config/core-pool-sysctl.yaml` for the sysctl system settings you added?

It'd be great if this PR can also include a new doc file that contains the command. As we discovered there are no docs for creating the core node pool.

Lastly maybe we can note what the GCE defaults are (or were) in the core-pool-sysctl.yaml file. For `tcp_mem` at least I think they were `359562	479418	719124`. It would just help to provide additional context to show how far off the path we're going. I'm guessing the defaults may change slightly over time, but we might say something like, "on YYYY-MM-DD the defaults were ..."

**Comment by shaneknapp** [comment]
> It's great this is being added!
> 
> Wouldn't the command you proposed [also need](https://cloud.google.com/kubernetes-engine/docs/how-to/node-system-config#create-node-pool) `--system-config-from-file=vendor/google/gke/node-pool/config/core-pool-sysctl.yaml` for the sysctl system settings you added?
> 
yes.  i remembered that i'd forgotten to add that, then forgot to remember to edit my comment.  :)

> It'd be great if this PR can also include a new doc file that contains the command. As we discovered there are no docs for creating the core node pool.
> 
yep, will do.  undecided on a README.md in that dir tree, but definitely in the docs/ dir.

> Lastly maybe we can note what the GCE defaults are (or were) in the core-pool-sysctl.yaml file. For `tcp_mem` at least I think they were `359562 479418 719124`. It would just help to provide additional context to show how far off the path we're going. I'm guessing the defaults may change slightly over time, but we might say something like, "on YYYY-MM-DD the defaults were ..."

sgtm, i'll make it so.

**Comment by shaneknapp** [comment]
ok, i found the page describing best (i believe) how to tune this crap:
https://cromwell-intl.com/open-source/performance-tuning/tcp.html

i'll make more changes and docs today.

**Comment by shaneknapp** [comment]
looks like we might be entering unsafe territory:
```
ERROR: (gcloud.container.node-pools.create) ResponseError: code=400, message=Unsupported kernel parameter net.ipv4.tcp_mem.
```

some random searching and i found this:  https://github.com/aws/karpenter/issues/1279

this led me to here:  https://kubernetes.github.io/ingress-nginx/examples/customization/sysctl/

investigating.

post-investigation:  that's just a command you run to patch a currently-running nginx node.  not what we're looking for.

**Comment by shaneknapp** [comment]
ok, i am punting on setting `net.ipv4.tcp_mem` for now as it's an 'unsafe' sysctl and it seems to be more involved to get that working than expected.

the list of 'safe' sysctls is here:  https://cloud.google.com/kubernetes-engine/docs/how-to/node-system-config#sysctl-options

i created a new node pool, and grabbed this from a node directly:
```
sknapp@gke-fall-2019-core-2023-04-18-8bc74a6a-nw1z ~ $ cat sysctls
net.core.netdev_max_backlog
net.core.rmem_max
net.core.wmem_max
net.ipv4.tcp_wmem
net.ipv4.tcp_rmem
net.ipv4.tcp_mem
sknapp@gke-fall-2019-core-2023-04-18-8bc74a6a-kj5f ~ $ for x in $(cat sysctls ); do grep $x /etc/sysctl.d/*; done
/etc/sysctl.d/99-gke-defaults.conf:net.core.netdev_max_backlog=1000
/etc/sysctl.d/99-gke-overrides.conf:net.core.netdev_max_backlog=30000
/etc/sysctl.d/99-gke-defaults.conf:net.core.rmem_max=212992
/etc/sysctl.d/99-gke-overrides.conf:net.core.rmem_max=3276800
/etc/sysctl.d/99-gke-defaults.conf:net.core.wmem_max=212992
/etc/sysctl.d/99-gke-overrides.conf:net.core.wmem_max=3276800
/etc/sysctl.d/99-gke-defaults.conf:net.ipv4.tcp_wmem=4096 16384 4194304
/etc/sysctl.d/99-gke-overrides.conf:net.ipv4.tcp_wmem=4096 87380 16777216
/etc/sysctl.d/99-gke-defaults.conf:net.ipv4.tcp_rmem=4096 87380 6291456
/etc/sysctl.d/99-gke-overrides.conf:net.ipv4.tcp_rmem=4096 87380 16777216
sknapp@gke-fall-2019-core-2023-04-18-8bc74a6a-kj5f ~ $ grep tcp_mem /etc/sysctl.d/*
sknapp@gke-fall-2019-core-2023-04-18-8bc74a6a-kj5f ~ $
```
confirmed that the 'overrides' are indeed overriding:
```
sknapp@gke-fall-2019-core-2023-04-18-8bc74a6a-kj5f ~ $ cat /proc/sys/net/ipv4/tcp_rmem
4096    87380   16777216
```
however...  `tcp_mem` is showing up here:
```
sknapp@gke-fall-2019-core-2023-04-18-8bc74a6a-kj5f ~ $ cat /proc/sys/net/ipv4/tcp_mem
769428  1025906 1538856
```
from [this blog](https://cromwell-intl.com/open-source/performance-tuning/tcp.html), he thoughtfully provides a black-magic formula:
```
# This server might have 200 clients simultaneously, so:
#   max(tcp_wmem) * 2 * 200 / 4096
net.ipv4.tcp_mem = 1638400 1638400 1638400
```
i did some 'maths'[tm], and `2 * 16777216 * 200 (random number pulled from that blog) / 4096 == 1638400`, which is super close to what `cat /proc/sys/net/ipv4/tcp_mem` shows as what's set for 'maximum':  `1538856`

this is MUCH higher when compared to the default settings in a pod (w/o this config):
```
net.ipv4.tcp_mem = 359562	479418	719124
```

TL;DR:  it looks like `tcp_mem` is magically being calculated at node/VM creation.  i think we're done.

edit:  ¯\\\_(ツ)\_/¯

**Comment by shaneknapp** [comment]
i'm going to merge this to staging...  it's also safe to merge to prod at any time!

File: vendor/google/gke/node-pool/config/core-pool-sysctl.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+   # net.core.netdev_max_backlog=1000
+   # net.core.rmem_max=212992
+   # net.core.wmem_max=212992
+   # net.ipv4.tcp_rmem=4096 87380 6291456
+   # net.ipv4.tcp_wmem=4096 16384 4194304
+   net.core.netdev_max_backlog: '30000'
+   net.core.rmem_max: '3276800'
+   net.core.wmem_max: '3276800'
+   net.ipv4.tcp_rmem: '4096 87380 16777216'
+   net.ipv4.tcp_wmem: '4096 87380 16777216'
```
Comment (consideRatio) [inline vendor/google/gke/node-pool/config/core-pool-sysctl.yaml:37]: @shaneknapp I appreciate these comments A LOT right now THANK YOU for writing them alongside this configuration!!

Comment (shaneknapp) [reply vendor/google/gke/node-pool/config/core-pool-sysctl.yaml:37]: thanks @consideRatio !

for some background (that i just realized was omitted from this PR and resides only in our JIRA), this was to help address our nginx ingress controller node running out of tcp kernel memory.  this would, in turn, start serving 500 and 503 to our users, connections would time out and we'd have to kill that pod and things would come back after a restart.

this would happen reliably ~30 days after instruction really kicked off, and due random factor of when, we almost never had a chance to do any forensics on the node.  and it would sometimes happen late in the semester but again, keep the ship afloat, etc etc.

this PR was just a patch to hopefully stop these failures from happening.  i wasn't hopeful but it should *help*...  i think?

in the end, it helped!  about 30 days after heavy usage, all the regular symptoms showed up and users started seeings 503s and timeouts.  about 45 minutes later, the core node was automatically killed and a fresh one cleared things up!  previously this could impacts us for hours, now minutes.

i also discovered that the tcp out of memory errors were showing up about ~30m before users became impacted.  we now have a page for this event, which should hopefully let us restart the node before everything goes down.  i'm also considering whipping up some monitoring for this and emitting to prom.

glad you liked the notes in the file.  when you get down in to kerneltuninglandia, there's not a lot of information out there (or a lot of 'just use these numbers' and 'hey!  it's fixed!').  i also knew  if i (or anyone else) needed to poke something this in the future there could at least be some breadcrumbs.

and to be completely fair:   this didn't fix anything, and while we're sure it's nginx, we haven't tried to debug that yet.  :)
